src/org/opensolaris/opengrok/analysis/StreamSource.java

```diff
@@ -62,6 +79,58 @@ public static StreamSource fromFile(final File file) {
             public InputStream getStream() throws IOException {
                 return new BufferedInputStream(new FileInputStream(file));
             }
+
+            @Override
+            public DigestedInputStream getSHA256stream() throws IOException {
+                return getSHA256stream(getStream());
+            }
         };
     }
+
+    protected static DigestedInputStream getSHA256stream(InputStream under)
+            throws IOException {
+        MessageDigest md;
+        try {
+            md = MessageDigest.getInstance(SHA256);
+        } catch (NoSuchAlgorithmException ex) {
+            /*
+             * NOT EXPECTED because "Every implementation of the Java
+             * platform is required to support ... SHA-256 ...."
+             */
+            throw new IOException("digest failed");
+        }
+        DigestInputStream dis = new DigestInputStream(under, md);
+
+        return new DigestedInputStream() {
+            @Override
+            public InputStream getStream() {
+                return dis;
+            }
+
+            @Override
+            public MessageDigest getMessageDigest() {
+                return dis.getMessageDigest();
+            }
+
+            @Override
+            public byte[] digestAll() throws IOException {
+                return StreamSource.digestAll(this);
+            }
+
+            @Override
+            public void close() throws Exception {
+                dis.close();
+            }
+        };
+    }
+
+    protected static byte[] digestAll(DigestedInputStream dis)
```

Comment:
I wonder if this should be part of DigestedInputStream (make it abstract class rather than interface) to avoid exposing MessageDigest in order to avoid anyone getting the value before the digest is fully computed.

Reply:
I argue no in my previous note.

Also, how would a third-party affirm that an `InputStream` is fully digested without possibly altering the stream unless a mediator was overlayed? That seems of dubious value.

Reply:
I was more after what happens if someone calls `getMessageDigest().digest()` before `digestAll()`. Seems like API deficiency to me.

Reply:
No consumer would call both of those because `getMessageDigest().digest()` is only used when there is a true reader of the data, and `digestAll()` is only used where the data is thrown away and only the digest is important.